autolab/Autolab
# LTI Roster Integration

## Description
Builds on the  [Implement LTI Advantage NRPS GET request PR](https://github.com/autolab/Autolab/pull/1649)
Implements logic to take the NRPS membership data and updates the roster of a course.

<img width="447" alt="Screen Shot 2022-11-28 at 12 03 55 AM" src="https://user-images.githubusercontent.com/11011325/204197911-228368f7-d645-4410-9ad5-d2d8c4989370.png">
Created new 'lti settings' modal.

<img width="1005" alt="Screen Shot 2022-11-28 at 12 04 32 AM" src="https://user-images.githubusercontent.com/11011325/204197985-e2a3555b-ab7a-45c5-b0b9-b4eaf6177f7f.png">
Created new 'sync course roster' page.

## How Has This Been Tested?
Going through the LTI linking instructions outlined in this [PR](https://github.com/autolab/Autolab/pull/1649)

Navigate to 'Manage Course Users'

**Test roster sync**
<img width="379" alt="Screen Shot 2022-11-28 at 12 05 23 AM" src="https://user-images.githubusercontent.com/11011325/204198070-86bc3b18-4594-445a-b11c-4917380d44c7.png">

 * Click the refresh button
<img width="1009" alt="Screen Shot 2022-11-28 at 12 08 12 AM" src="https://user-images.githubusercontent.com/11011325/204198404-a6fca7e5-1a71-41f2-adb6-382c28d696da.png">

 * Press confirm
 * Verify that additional students have been added to the course.

**Test dropping missing students**
 * Click 'Linked Course Settings' on the 'Manage Course Users' page
<img width="394" alt="Screen Shot 2022-11-28 at 12 09 52 AM" src="https://user-images.githubusercontent.com/11011325/204198615-ffee78bd-9fc4-4aca-86ca-e21d0de449cd.png">

 * Check 'Auto drop students not enrolled in linked course', then click 'Save'
 * Ensure the course has students enrolled that don't have email *@example.org (i.e. students manually added or imported via a roster)
<img width="379" alt="Screen Shot 2022-11-28 at 12 05 23 AM" src="https://user-images.githubusercontent.com/11011325/204198070-86bc3b18-4594-445a-b11c-4917380d44c7.png">

 * Click the refresh button
<img width="966" alt="Screen Shot 2022-11-28 at 12 15 04 AM" src="https://user-images.githubusercontent.com/11011325/204199199-d2489191-7d96-4b4e-9fc6-3a3273270648.png">

 * Verify that students missing in the NRPS roster (non @example.org emails) are red
 * Click 'Confirm' and verify that the students that appeared red are now dropped (via the dropped column on the 'Manage Course Users' page)

**Test unlinking a course**
 * Click 'Linked Course Settings'
<img width="440" alt="Screen Shot 2022-11-28 at 12 17 09 AM" src="https://user-images.githubusercontent.com/11011325/204199473-caaa7998-0944-4854-a53e-8ddf8948ae37.png">

 * Click 'unlink' and verify a confirmation window pops up, click 'OK'
<img width="238" alt="Screen Shot 2022-11-28 at 12 17 59 AM" src="https://user-images.githubusercontent.com/11011325/204199563-73c9f550-bfe8-46d4-9723-49b0a21789e6.png">

 * Verify the course was unlinked (via success flash, the 'Linked Course Settings' button, and 'Roster last synced with linked course: ...' text should no longer appear on the 'Manage Course Users' page)



## Types of changes


## Checklist:
